src/main/scala/sample/stream/GroupLogFile.scala

```diff
@@ -25,21 +25,25 @@ object GroupLogFile {
     // read lines from a log file
     val logFile = new File("src/main/resources/logfile.txt")
 
-    import akka.stream.io.Implicits._ // add file sources to Source object or use explicitly: SynchronousFileSource(f)
-    Source.synchronousFile(logFile).
+    FileIO.fromFile(logFile).
       // parse chunks of bytes into lines
       via(Framing.delimiter(ByteString(System.lineSeparator), maximumFrameLength = 512, allowTruncation = true)).
       map(_.utf8String).
+      map {
+        case line@LoglevelPattern(level) => (level, line)
+        case line@other => ("OTHER", line)
+      }.
       // group them by log level
-      groupBy {
-        case LoglevelPattern(level) => level
-        case other                  => "OTHER"
+      groupBy(5, _._1).
+      fold(("", List.empty[String])) {
+        case ((_, list), (level, line)) => (level, line :: list)
       }.
       // write lines of each group to a separate file
       mapAsync(parallelism = 5) {
-        case (level, groupFlow) =>
-          groupFlow.map(line => ByteString(line + "\n")).runWith(Sink.synchronousFile(new File(s"target/log-$level.txt")))
+        case (level, groupList) =>
+          Source(groupList.reverse).map(line => ByteString(line + "\n")).runWith(FileIO.toFile(new File(s"target/log-$level.txt")))
```

Comment:
The group key is part of every element, otherwise it could not have been extracted by groupBy. When it comes to exposing a file sink that takes the filename from one part of the element while taking the bytes from another there are multiple options which I currently have no time to investigate.